### PR TITLE
!!! TASK: Workspace aware commands

### LINKED ISSUE #4694: Refer to `WorkspaceName` instead of `ContentStreamId` in commands
Currently, most commands (all implementing the `RebasableToOtherContentStreamsInterface`) refer to the *content stream id*.
This has several issues:

* Content Stream are mostly an _implementation detail_. The _intention_ is to commit changes to a **workspace**
* As a result, the common use case is cumbersome right now (fetch workspace via `WorkspaceFinder` to determine content stream id, ...)
* When we introduce "command interception" for privilege enforcement (see https://github.com/neos/neos-development-collection/pull/4185#issuecomment-1777399632) it will be harder to determine the allowed content streams for the authenticated user
* To re-apply commands during rebasing, we need to change the CS id currently (`createCopyForContentStream()`) – by referring to the workspace, we should be able to simply copy the commands
* Lastly, by referring to the workspace, we will be able to implement some retry mechanism when aggregate checks fail due to a changed content stream during publish/rebase (see #4680 for example)

### PR DESCRIPTION
**Pull request by nezaniel**
This makes commands operate on workspaces instead of content streams, as this
* better reflects the User's intention
* gives the command handlers options to remediate merge conflicts, e.g when the workspace has been rebased in the meantime and its previous content stream no longer exists

Notice: For Rebasing, a surrogate WorkspaceWasRebased is used. This passes the tests but does not make much sense, instead we should introduce a new event `WorkspaceRebaseWasStarted` that sets the workspace to the new content stream so that during rebase the commands can be applied to the proper content stream.

Resolves: #4694

Neos Ui Part: https://github.com/neos/neos-ui/pull/3661

**Upgrade instructions**

The php api of most content repository commands was adjusted. The content stream id argument was removed and replaced with the workspace name.
(To migrate you can also look into the Ui [adjustments](https://github.com/neos/neos-ui/pull/3661) for inspiration)

```php
$contentRepository->handle(
    SetNodeProperties::create(
        WorkspaceName::forLive(), // instead of $contentStreamId
        $subject->nodeAggregateId,
        $originDimensionSpacePoint,
        PropertyValuesToWrite::fromArray(['foo' => 'bar'])
    )
);
```

> [!IMPORTANT]  
> This change is partially breaking to your existing events. To allow discarding or publishing a workspace the migration has to be applied:
> ```sh
> flow migrateevents:migrateMetaDataToWorkspaceName
> ```

**Checklist**

- [x] Code follows the PSR-12 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the 9.0 branch
- [x] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [x] Reviewer - The first section explains the change briefly for change-logs
- [x] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

**Comment by kitsunet** [comment]
So much work with all the tests, but this looks great!

**Review by bwaidelich — COMMENTED** [review]
Amazing, thanks for your hard work!

Can you give a bit more context to your notice re "For Rebasing, a surrogate WorkspaceWasRebased is used."?

**Comment by nezaniel** [comment]
> Amazing, thanks for your hard work!
> 
> Can you give a bit more context to your notice re "For Rebasing, a surrogate WorkspaceWasRebased is used."?

Sure.
When rebasing (see https://github.com/neos/neos-development-collection/pull/4708/files#diff-3e5de1a138586c8c85cdd71ed1a765ecc0ba0d79ab2579a6000475d427cefda2L391), we forked a content stream, modified the commands to operate on that content stream and then applied them.

This no longer works since the commands now operate on workspaces and the workspace doesn't change during rebase. Thus, before the commands get applied, the workspace must already point to the new content stream. This can currently only be done by `WorkspaceWasRebased`, which is semantically wrong and should be replaced with something like `WorkspaceRebaseWasStarted`

**Comment by mhsdesign** [comment]
So okay i ran the neos ui e2e test locally in combination with https://github.com/neos/neos-ui/pull/3661 and they pass.

At first i had little trouble getting this branch to work and emptied my db:

> Warning: Undefined array key "workspaceName" in Neos.ContentRepository.Core/Classes/Feature/WorkspacePublication/Dto/NodeIdToPublishOrDiscard.php line 48

This WorkspaceWasPartiallyDiscarded event could not be deserialized when catching up:

```
{
    "workspaceName": "user-admin",
    "newContentStreamId": "ed1314b0-93cc-4bf1-b249-48e5f0c0449e",
    "previousContentStreamId": "543da977-ad93-4785-ae6c-b68333b16f8c",
    "discardedNodes": [
        {
            "contentStreamId": "543da977-ad93-4785-ae6c-b68333b16f8c",
            "nodeAggregateId": "5b0d6ac0-40ab-47e8-b79e-39de6c0700df",
            "dimensionSpacePoint": {
                "language": "en_US"
            }
        },
  ...
```

i have to test again that this change is really not breaking if swapped out on the fly

**Comment by kitsunet** [comment]
@mhsdesign did you retest this?

**Comment by kitsunet** [comment]
I tried it as well and so far didn't notice problems, so I would realyl like to get this in ASAP.

**Review by kitsunet — APPROVED** [review]

**Comment by ahaeslich** [comment]
> At first i had little trouble getting this branch to work and emptied my db:

@kitsunet so you didn't ran into this error?

> This passes the tests but does not make much sense, instead we should introduce a new event WorkspaceRebaseWasStarted that sets the workspace to the new content stream so that during rebase the commands can be applied to the proper content stream.

So this should be added in a follow up or do we need to introduce this event here? If we can add this later I'm with you on merging this ASAP.

**Comment by kitsunet** [comment]
Seemed fine for me, but my dataset is small. I applied patch, did cr:setup on my exisating database for good measure and tried the website.

**Comment by mhsdesign** [comment]
Before merging this there should be some things about the changed publishing be discussed. Im not really sure which places @nezaniel meant, but maybe someone can answer the open points in the next weekly ;)

**Comment by mhsdesign** [comment]
> did you retest this?

now i did and i ran into the same error. Neither `flow cr:setup` or `flow cr:replayall` can fix the problem. Only a complete reset and reimport.

Maybe you can reproduce this also on your machines otherwise im happy to pair debug my case (didt got far alone)

on neos 9-dev

import a site (in my case the neos ui e2e testing site)

switch to your branch and create a content node.

Edit: for some reason the neos demo is not affected... hmm

`structureadjustments:detect` didnt found anything

File: Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php]: As discussed earlier today:
I think it is very dangerous to change the order of events because it will lead to a point in time where the workspace points to a broken content stream etc.

So one pragmatic solution we came up with was to temporarily change the workspace-to-contentstream resolution while re-applying commands.
AFAIS this is always done in the `ConstraintChecks` trait in `requireContentStream()`, so maybe we could add some (very internal) closure to that trait along the lines of:

```php
private ?ContentStreamId $_overriddenContentStreamid = null;
public function withContentStreamId(ContentStreamId $contentStreamId, \Closure $closure): void {
    $previousValue = $this->_overriddenContentStreamid;
    $this->_overriddenContentStreamid = $contentStreamId;
    try {
        $closure();
    } finally {
    $this->_overriddenContentStreamid = $previousValue;
    }
  }
```

and then in `requireContentStream()` add
```php
        $contentStreamId = $this->_overriddenContentStreamid ?? $contentRepository->getWorkspaceFinder()->findOneByName($workspaceName)
            ?->currentContentStreamId;
```

And use it below like:

```php
$this->withContentStreamId($newContentStreamId, function () {
  $originalCommands = $this->extractCommandsFromContentStreamMetadata($workspaceContentStreamName)
  $this->applyOriginalCommands($originalCommands);
});
```

**Comment by mhsdesign** [comment]
~The feature NodeTraversal-DescendantNodes in `Neos.ContentRepository.BehavioralTests` seems to be flaky on this branch.~

Not only here but also on dev-9.0 -> new issue for this separate topic: https://github.com/neos/neos-development-collection/issues/4774

**Comment by mhsdesign** [comment]
Okay. I retested again. And found a reproducible bug https://github.com/neos/neos-development-collection/pull/4708#issuecomment-1812671866. If there are pending changes in a workspace the reabase will fail (and thus prevent a login) even discard via cli is not possible.

Additionally the discarding from the publish menu is not correctly reflected in the neos ui - but this is a temporary display glitch

**Comment by mhsdesign** [comment]
> Additionally the discarding from the publish menu is not correctly reflected in the neos ui - but this is a temporary display glitch

I dont know why its happening but i at least can now categorise the problem: A timing issue :/

The UI's [`workspaceService->getPublishableNodeInfo`](https://github.com/neos/neos-ui/blob/24daee6e4df89152cf34f7527f8f62be30bcdbd8/Classes/Domain/Model/Feedback/Operations/UpdateWorkspaceInfo.php#L124) doesnt work immediately after BLOCKED handling a cr publish anymore. A sleep statement of 1 second fixed this in my testings _or_ making the `publishAction` or `discardAction` synchronous via `CatchUpTriggerWithSynchronousOption::enableSynchronicityForSpeedingUpTesting()`

In sync, the ui e2e pass and the ui works.

But 🤔 previously i _said_ that the e2e tests work ... so maybe we introduced a regression since november and this pr initially worked? https://github.com/neos/neos-development-collection/pull/4708#issuecomment-1812671866

-------

The above mentioned [error](https://github.com/neos/neos-development-collection/pull/4708#issuecomment-1812671866) will also show up after a first login but when doing a relogin again it works. Seems that the initial rebase has a hickup?

> Warning: Undefined array key "workspaceName" in /Neos/Neos.ContentRepository.Core/Classes/Feature/NodeVariation/Command/CreateNodeVariant.php line 70

File: Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php]: @nezaniel i think you have some local changes lying around that fix this? 

Feel free to force push, i can cherry-pick my only _real_ change on top: https://github.com/neos/neos-development-collection/pull/4708/commits/7a4877caa83855054fd3cefb51de44ffc8e3328f

File: Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php]:
```suggestion
            // We don't need to adjust commands to the new content stream (createCopyForContentStream) as the workspace stays the same
```

File: Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php]: TODO over here.

File: Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php:291]: we dont do that anymore correct?

edit, seems still correct.

**Comment by mhsdesign** [comment]
Hmm after importing the demo on a fresh 9.0, changing the title and publishing and switching to this branch a full `cr:replayAll` will fail.

> Warning: Undefined array key "workspaceName" in /Neos.ContentRepository.Core/Classes/Feature/WorkspacePublication/Dto/NodeIdToPublishOrDiscard.php line 46

that seems to be because the stored command in the events need to be migrated?

interestingly `cr:replayAll` works with a fresh import of the neos demo ... that seems to correlate with the fact that the imported events dont have a `commandClass` in the payload????????????????????????????????????????????????

**Review by mhsdesign — CHANGES_REQUESTED** [review]
... because we should not merge this if one needs a clean db imo ... at least not if there is no way around this ...

**Comment by mhsdesign** [comment]
The replay issue also happens when i previously try to prune everything outdated:

First import the Neos.Demo at 9.0-dev then create a new user workspace (by logging in) then make changes via `NodePropertiesWereSet` e.g.)

And publish these changes.

Clean stuff:

```
flow contentstream:prune
flow contentstream:pruneremovedfromeventstream
```


Switch to this branch do a `cr:setup`.

And the `flow cr:replayAll` will fail:

```
Replaying events for all projections of Content Repository "default" ...
Warning: Undefined array key "workspaceName" in /Neos.ContentRepository.Core/Classes/Feature/WorkspacePublication/Dto/NodeIdToPublishOrDiscard.php line 46
```

the last 5 events:
```
218,ContentStream:d099734d-ea67-432e-b003-ab664f978421,1,NodePropertiesWereSet,"{""contentStreamId"":""d099734d-ea67-432e-b003-ab664f978421"",""nodeAggregateId"":""d17caff2-f50c-d30b-b735-9b9216de02e9"",""originDimensionSpacePoint"":{""language"":""en_US""},""affectedDimensionSpacePoints"":[{""language"":""en_US""}],""propertyValues"":{""title"":{""value"":""Hakk"",""type"":""string""}}}","{""commandClass"":""Neos\\ContentRepository\\Core\\Feature\\NodeModification\\Command\\SetSerializedNodeProperties"",""commandPayload"":{""contentStreamId"":""d099734d-ea67-432e-b003-ab664f978421"",""nodeAggregateId"":""d17caff2-f50c-d30b-b735-9b9216de02e9"",""originDimensionSpacePoint"":{""language"":""en_US""},""propertyValues"":{""title"":{""value"":""Hakk"",""type"":""string""}}},""initiatingUserId"":""3661a51c-e94f-4023-b5c3-a1ab9d63bce4"",""initiatingTimestamp"":""2024-01-31T13:55:08+00:00""}"
219,ContentStream:abcec354-2049-487b-8734-5a96cd9fc151,0,ContentStreamWasForked,"{""newContentStreamId"":""abcec354-2049-487b-8734-5a96cd9fc151"",""sourceContentStreamId"":""d099734d-ea67-432e-b003-ab664f978421"",""versionOfSourceContentStream"":1}","{""initiatingUserId"":""3661a51c-e94f-4023-b5c3-a1ab9d63bce4"",""initiatingTimestamp"":""2024-01-31T13:55:08+00:00""}"
220,ContentStream:812189a9-b523-4702-9e52-735bb5f28d1c,212,NodePropertiesWereSet,"{""contentStreamId"":""812189a9-b523-4702-9e52-735bb5f28d1c"",""nodeAggregateId"":""d17caff2-f50c-d30b-b735-9b9216de02e9"",""originDimensionSpacePoint"":{""language"":""en_US""},""affectedDimensionSpacePoints"":[{""language"":""en_US""}],""propertyValues"":{""title"":{""value"":""Hakk"",""type"":""string""}}}","{""commandClass"":""Neos\\ContentRepository\\Core\\Feature\\NodeModification\\Command\\SetSerializedNodeProperties"",""commandPayload"":{""contentStreamId"":""d099734d-ea67-432e-b003-ab664f978421"",""nodeAggregateId"":""d17caff2-f50c-d30b-b735-9b9216de02e9"",""originDimensionSpacePoint"":{""language"":""en_US""},""propertyValues"":{""title"":{""value"":""Hakk"",""type"":""string""}}},""initiatingUserId"":""3661a51c-e94f-4023-b5c3-a1ab9d63bce4"",""initiatingTimestamp"":""2024-01-31T13:55:08+00:00""}"
221,ContentStream:d099734d-ea67-432e-b003-ab664f978421,2,ContentStreamWasRemoved,"{""contentStreamId"":""d099734d-ea67-432e-b003-ab664f978421""}","{""initiatingUserId"":""3661a51c-e94f-4023-b5c3-a1ab9d63bce4"",""initiatingTimestamp"":""2024-01-31T13:55:09+00:00""}"
222,Workspace:user-admin,1,WorkspaceWasPartiallyPublished,"{""sourceWorkspaceName"":""user-admin"",""targetWorkspaceName"":""live"",""newSourceContentStreamId"":""abcec354-2049-487b-8734-5a96cd9fc151"",""previousSourceContentStreamId"":""fa799f20-bbb4-4aa1-a909-c8f1329bc247"",""publishedNodes"":[{""contentStreamId"":""fa799f20-bbb4-4aa1-a909-c8f1329bc247"",""nodeAggregateId"":""d17caff2-f50c-d30b-b735-9b9216de02e9"",""dimensionSpacePoint"":{""language"":""en_US""}}]}","{""initiatingUserId"":""3661a51c-e94f-4023-b5c3-a1ab9d63bce4"",""initiatingTimestamp"":""2024-01-31T13:55:09+00:00""}"
```

**Comment by mhsdesign** [comment]
As discussed we should write a command (separate pr targeting the `workspaceAwareCommands` branch) to migrate the events table.

In the meta data table the `commandClass`'s `commandPayload` must be changed from providing a `contentStreamId` to additionally have a `workspaceName` field. We should be able to use the workspace finder by content stream for this, but in case no workspace is resolve able a dummy value must be set to allow the deserialisation.

File: Neos.ContentRepository.Core/Classes/ContentRepository.php
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/ContentRepository.php]: todo might be obsolete due to markStale interface and auto invalidation

File: Neos.ContentRepository.Core/Classes/Projection/Workspace/WorkspaceFinder.php
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Projection/Workspace/WorkspaceFinder.php]: todo why?

File: Neos.ContentRepository.Core/Classes/ContentRepository.php
Comment (nezaniel) [reply Neos.ContentRepository.Core/Classes/ContentRepository.php]: obsolete and removed

File: Neos.ContentRepository.Core/Classes/Projection/Workspace/WorkspaceFinder.php
Comment (nezaniel) [reply Neos.ContentRepository.Core/Classes/Projection/Workspace/WorkspaceFinder.php]: obsolete and removed

File: Neos.ContentRepository.BehavioralTests/Tests/Functional/Feature/WorkspacePublication/WorkspaceWritingDuringPublication.php
Comment (bwaidelich) [inline Neos.ContentRepository.BehavioralTests/Tests/Functional/Feature/WorkspacePublication/WorkspaceWritingDuringPublication.php]:
```suggestion
    public function thenConcurrentCommandsLeadToAnException(): void
```

**Review by dlubitz — APPROVED** [review]
Thank you very much for all the work!

**Comment by mhsdesign** [comment]
1. the event migration `flow migrateevents:fillWorkspaceNameInCommandPayloadOfEventMetaData` is ready to merged here into it https://github.com/neos/neos-development-collection/pull/4890

2. in the pr i tried to make the override cs stream flow more obvious, but its just a personal nitpick

3. im not sure if we have to adjust any inline code documentation about the publishing, i could only find these sketches which dont seem to be affected so all good ;)

```
./Neos.ContentRepository.Core/Classes/Feature/NodeMove/Event/NodeAggregateWasMoved.monopic
./Neos.ContentRepository.Core/Classes/Projection/ContentStream/ContentStreamFinder.monopic
./Neos.Neos/Classes/Fusion/Cache/GraphProjectorCatchUpHookForCacheFlushing.monopic
./Neos.Neos/Classes/FrontendRouting/EventSourcedFrontendNodeRoutePartHandler.monopic
```

4. ~i cannot reproduce reliable the neos ui e2e failure documented here https://github.com/neos/neos-development-collection/pull/4708#issuecomment-1917846367 so this should not be a blocker and might be another problem that existed beforehand.~
edit2: yes i can reproduce it, i had a `sleep(1)` statement in the code
edit3: the problem was fixed with the merge of `4583-variantRecreation` because the code was refactored and contained previously unsafe `publishEvents` which was not blocked in the `handleDiscardIndividualNodesFromWorkspace` in this line https://github.com/neos/neos-development-collection/blob/b194763769dc36169e61f4fc8752ad3247fbeb96/Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php#L637-L650

-> this was fixed

5. ~now the ui is failing at this moment with ~
> Content stream "766799e1-2a06-4b15-8986-e156e9d74b60" is closed.

-> this was fixed

**Review by mhsdesign — APPROVED** [review]
Thank you so much for taking care.

**Comment by bwaidelich** [comment]
Yeah!! Thanks so much, @nezaniel and congratulations!!

**Pull request merged**
